src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -150,6 +161,64 @@ private void handleSetKeyset(RoutingContext rc) {
         }
     }
 
+    // Method to check if one set contains any values from another set
+    private static <T> boolean containsAny(Set<T> set1, Set<T> set2) {
+        for (T element : set2) {
+            if (set1.contains(element)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void handleListAllKeysetsRelated(RoutingContext rc) {
+        try {
+            // Get value for site id
+            final Optional<Integer> siteIdOpt = RequestUtil.getSiteId(rc, "site_id");
+            if (!siteIdOpt.isPresent()) return;
+            final int siteId = siteIdOpt.get();
+
+            if (siteId != Const.Data.AdvertisingTokenSiteId && !SiteUtil.isValidSiteId(siteId)) {
+                ResponseUtil.error(rc, 400, "must specify a valid site id");
+                return;
+            }
+
+            // Get value for client type from the backend
+            Set<ClientType> clientTypes = this.siteProvider.getSite(siteId).getClientTypes();
+
+            // Check if this site has any client key that has an ID_READER role
+            boolean isIdReaderRole = false;
+            for (LegacyClientKey c : this.clientKeyProvider.getAll()) {
+                if (c.getRoles().contains(Role.ID_READER)) {
+                    isIdReaderRole = true;
+                }
+            }
+
+            // Get the keyset ids that need to be rotated
+            final JsonArray ja = new JsonArray();
+            Map<Integer, AdminKeyset> collection = this.keysetProvider.getSnapshot().getAllKeysets();
+            for (Map.Entry<Integer, AdminKeyset> keyset : collection.entrySet()) {
+                // The keysets meet any of the below conditions ALL need to be rotated:
+                // a. Keysets where allowed_types include any of the clientTypes of the site
+                // b. If this participant has a client key with ID_READER role, we want to rotate all the keysets where allowed_sites is set to null
+                // c. Keysets where allowed_sites include the leaked site
+                // d. Keysets belonging to the leaked site itself
+                if (containsAny(keyset.getValue().getAllowedTypes(), clientTypes) ||
```

Comment:
Btw there is a method [canClientAccessKey](https://github.com/IABTechLab/uid2-shared/blob/19edb010c6a4d753d03c89268c238be10a8f6722/src/main/java/com/uid2/shared/auth/KeysetSnapshot.java#L13) which is the source of truth. 

Would be great if we could use that here, although might not be feasible. If we can't use it, still worth mentioning it here as a comment.